utils/utils.py

```diff
@@ -418,3 +420,192 @@ def save_notified_events(notified_events: Dict[str, Any]) -> None:
+                contributor.emoji_id = emoji_id
+            else:
+                contributor = Contributor(
+                    uid=uid, note=note, server_name=str(guild_id), emoji_id=emoji_id
+                )
+                session.add(contributor)
+
+            session.commit()
+
+    @staticmethod
+    def get_contributors_from_db(guild_id: int) -> List[Contributor]:
+        """Get list of contributors for a specific server"""
+        with SessionLocal() as session:
+            contributors = session.query(Contributor).filter_by(server_name=str(guild_id)).all()
+            return contributors
+
+    @staticmethod
+    def save_event(
+        event_id: int,
+        guild_id: int,
+        posted_at: Optional[float] = None,
+        notified_at: Optional[float] = None,
+    ) -> None:
+        """Save or update an event in the database"""
+        with SessionLocal() as session:
+            event = session.query(Event).filter_by(event_id=event_id).first()
+            if event:
+                if posted_at is not None:
+                    event.posted_at = posted_at
+                if notified_at is not None:
+                    event.notified_at = notified_at
+            else:
+                event = Event(
+                    event_id=event_id,
+                    guild_id=guild_id,
+                    posted_at=posted_at,
+                    notified_at=notified_at,
+                )
+                session.add(event)
+            session.commit()
+
+    @staticmethod
+    def get_posted_events() -> List[int]:
+        """Get list of posted event IDs from database"""
+        with SessionLocal() as session:
+            events = session.query(Event).filter(Event.posted_at.isnot(None)).all()
+            return [event.event_id for event in events]
+
+    @staticmethod
+    def get_notified_events() -> Dict[int, float]:
+        """Get dictionary of notified events with their notification timestamps"""
+        with SessionLocal() as session:
+            events = session.query(Event).filter(Event.notified_at.isnot(None)).all()
+            return {event.event_id: event.notified_at for event in events}
+
+    @staticmethod
+    def save_ongoing_vote(proposal_id: str, vote_data: dict):
+        """Save an ongoing vote to the database"""
+        with SessionLocal() as session:
+            vote = (
+                session.query(OngoingVote)
+                .filter(OngoingVote.proposal_id == proposal_id)
+                .first()
+            )
+            if not vote:
+                vote = OngoingVote(proposal_id=proposal_id)
+
+            vote.draft = vote_data.get("draft", {})
+            vote.end_time = vote_data.get("end_time", 0)
+            vote.title = vote_data.get("title", "")
+            vote.channel_id = vote_data.get("channel_id", "")
+            vote.thread_id = vote_data.get("thread_id", "")
+            vote.message_id = vote_data.get("message_id", "")
+
+            session.add(vote)
```

Comment:
does this not need to be conditional on `if not vote`? what happens if you "add" a vote that already exists?